presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedInPredicateToJoin.java

```diff
@@ -0,0 +1,431 @@
+    }
+
+    private Optional<PlanNode> apply(
+            ApplyNode apply,
+            InPredicate inPredicate,
+            Symbol inPredicateOutputSymbol,
+            Lookup lookup,
+            PlanNodeIdAllocator idAllocator,
+            SymbolAllocator symbolAllocator)
+    {
+        Optional<Decorrelated> decorrelated = new DecorrelatingVisitor(lookup, apply.getCorrelation())
+                .visit(apply.getSubquery());
+
+        if (!decorrelated.isPresent()) {
+            return Optional.empty();
+        }
+
+        PlanNode projection = buildInPredicateEquivalent(
+                apply,
+                inPredicate,
+                inPredicateOutputSymbol,
+                decorrelated.get(),
+                idAllocator,
+                symbolAllocator
+        );
+
+        return Optional.of(projection);
+    }
+
+    private PlanNode buildInPredicateEquivalent(
+            ApplyNode apply,
+            InPredicate inPredicate,
+            Symbol inPredicateOutputSymbol,
+            Decorrelated decorrelated,
+            PlanNodeIdAllocator idAllocator,
+            SymbolAllocator symbolAllocator)
+    {
+        Expression correlationCondition = and(decorrelated.getCorrelatedPredicates());
+        PlanNode decorrelatedBuildSource = decorrelated.getNode();
+
+        AssignUniqueId probeSide = new AssignUniqueId(
+                idAllocator.getNextId(),
+                apply.getInput(),
+                symbolAllocator.newSymbol("unique", BIGINT));
+
+        Symbol buildSideKnownNonNull = symbolAllocator.newSymbol("buildSideKnownNonNull", BIGINT);
+        ProjectNode buildSide = new ProjectNode(
+                idAllocator.getNextId(),
+                decorrelatedBuildSource,
+                Assignments.builder()
+                        .putAll(Assignments.identity(decorrelatedBuildSource.getOutputSymbols()))
+                        .put(buildSideKnownNonNull, bigint(0))
+                        .build()
+        );
+
+        Symbol probeSideSymbol = Symbol.from(inPredicate.getValue());
+        Symbol buildSideSymbol = Symbol.from(inPredicate.getValueList());
+
+        Expression joinExpression = and(
+                or(
+                        new IsNullPredicate(probeSideSymbol.toSymbolReference()),
+                        new ComparisonExpression(ComparisonExpressionType.EQUAL, probeSideSymbol.toSymbolReference(), buildSideSymbol.toSymbolReference()),
+                        new IsNullPredicate(buildSideSymbol.toSymbolReference())
+                ),
+                correlationCondition
+        );
+
+        JoinNode leftOuterJoin = leftOuterJoin(idAllocator, probeSide, buildSide, joinExpression);
+
+        Expression matchCondition = and(
+                isNotNull(probeSideSymbol),
+                isNotNull(buildSideSymbol)
+        );
+        Entry<Symbol, AggregationNode.Aggregation> countMatches = countWithFilter(symbolAllocator, "countMatches", matchCondition);
```

Comment:
maybe allocate symbols for `countMatches` and `countNullMatches` here (outside of `countWithFilter`), pass those symbols as a parameters to `buildAggregationNode` helper method that would encapsulate lines 197 to 221, and use those allocated symbols in the lines 224 and 225 directly (instead of `countMatches.getKey()`)?

Together with above comment that would simplify this method to something like:

```
Symbol probeSideSymbol = ...
Symbol buildSideSymbol = ...

JoinNode leftOuterJoin = buildJoin(...);

Symbol countMatches = ...
Symbol countNullMatches = ...

Aggregate aggregation = buildAggragation(...)

return buildProjection(....);
```